File: app/schemas/expenses.py
"""Unified expense schemas – merge receipts + email expenses."""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class UnifiedExpenseResponse(BaseModel):
    """Single expense from either receipt or Gmail."""
    id: str
    source: str  # "receipt" or "gmail"
    source_id: str  # receipt_id or gmail_message_id
    merchant: str
    normalized_merchant: str
    amount: float
    currency: str = "USD"
    transaction_at: Optional[datetime] = None
    category: Optional[str] = None
    payment_method: Optional[str] = None
    expense_type: Optional[str] = None
    confidence: float = 0.0
    meta: dict[str, Any] = Field(default_factory=dict)


class UnifiedExpenseSummaryResponse(BaseModel):
    """Combined spend summary across receipts and emails."""
    total_today: float = 0.0
    total_this_month: float = 0.0
    by_source: dict[str, float] = Field(default_factory=dict)
    by_category: list[dict] = Field(default_factory=list)
    recent_merchants: list[str] = Field(default_factory=list)
    recent_expenses: list[UnifiedExpenseResponse] = Field(default_factory=list)

File: frontend/src/api/expenses.ts
import client from './client';

export interface UnifiedExpense {
  id: string;
  source: 'receipt' | 'gmail';
  source_id: string;
  merchant: string;
  normalized_merchant: string;
  amount: number;
  currency: string;
  transaction_at: string;
  category: string;
  payment_method?: string;
  expense_type?: string;
  confidence: number;
  meta?: {
    subject?: string;
    store_name?: string;
    receipt_id?: string;
    snippet?: string;
  };
}

export interface UnifiedExpenseList {
  items: UnifiedExpense[];
  total: number;
}

export interface UnifiedExpenseSummary {
  today_total: number;
  month_total: number;
  by_source: { receipt: number; gmail: number };
  by_category: Array<{ category: string; total: number }>;
  recent_merchants: string[];
}

export interface EmailExpense {
  id: string;
  user_id: string;
  gmail_message_id: string;
  source: 'gmail';
  sender: string;
  subject: string;
  snippet: string;
  received_at: string;
  expense_type: string;
  merchant: string;
  normalized_merchant: string;
  amount: number;
  currency: string;
  tax?: number;
  transaction_at: string;
  category: string;
  payment_method?: string;
  confidence: number;
  parsing_status: string;
  review: {
    needsHumanReview: boolean;
    reviewed: boolean;
    reviewedAt?: string;
  };
  raw_text: string;
  created_at: string;
}

export const expensesApi = {
  list: (params?: {
    source?: string;
    category?: string;
    from_date?: string;
    to_date?: string;
    merchant?: string;
    limit?: number;
    offset?: number;
  }) => client.get<never, UnifiedExpenseList>('/api/v1/expenses', { params }),

  summary: () =>
    client.get<never, UnifiedExpenseSummary>('/api/v1/expenses/summary'),

  emailExpenses: (params?: { page?: number; page_size?: number }) =>
    client.get<never, { items: EmailExpense[]; pagination: { page: number; page_size: number; total: number; total_pages: number } }>('/api/v1/email-expenses', { params }),

  reviewEmailExpense: (id: string, data: {
    merchant?: string;
    amount?: number;
    currency?: string;
    category?: string;
    expense_type?: string;
    transaction_at?: string;
  }) => client.patch<never, EmailExpense>(`/api/v1/email-expenses/${id}/review`, data),
};
